components/list/src/provider/mod.rs

```diff
@@ -74,31 +73,36 @@ pub const MARKERS: &[DataMarkerInfo] = &[
     derive(serde::Serialize, databake::Bake),
     databake(path = icu_list::provider),
 )]
-pub struct ListFormatterPatternsV2<'data>(
-    #[cfg_attr(feature = "datagen", serde(with = "deduplicating_array"))]
-    /// The patterns in the order start, middle, end, pair, short_start, short_middle,
-    /// short_end, short_pair, narrow_start, narrow_middle, narrow_end, narrow_pair,
-    pub [ConditionalListJoinerPattern<'data>; 12],
-);
-
-#[cfg(feature = "serde")]
-impl<'de> serde::Deserialize<'de> for ListFormatterPatternsV2<'de> {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::de::Deserializer<'de>,
-    {
-        #[cfg(not(feature = "serde_human"))]
-        if deserializer.is_human_readable() {
-            use serde::de::Error;
-            return Err(D::Error::custom(
-                    "Deserializing human-readable ListFormatter data requires the 'serde_human' feature",
-                ));
-        }
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct ListFormatterPatternsV2<'data> {
+    /// The start pattern
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub start: ListJoinerPattern<'data>,
+    /// The middle pattern. It doesn't need to be a pattern because it has to start with `{0}`
+    /// and end with `{1}`, so we just store the string in between.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub middle: Cow<'data, str>,
+    /// The end pattern
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub end: ConditionalListJoinerPattern<'data>,
+    /// The pair pattern, if it's different from the end pattern.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub pair: Option<ConditionalListJoinerPattern<'data>>,
+}
```

Comment:
Thought: we might consider changing this to use `Pattern`

Reply:
I tried, pattern needs work to be usable, and it doesn't save much compared to this